[1단계 - 페이먼츠 모듈] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
안녕하세요, 마르코! 니야입니다 😄
'점심 뭐 먹지' 미션에 이어 다시 뵙게 되어 반갑습니다!
지난 미션 때 많은 도움을 받아 정말 만족스러웠는데, 이번 페이먼츠 모듈 미션도 잘 부탁드리겠습니다 🙇🏻‍♀️

# 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

## 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

# 🕵️ 셀프 리뷰(Self-Review)

## 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - `components` npm 배포 링크: [@sooyeoniya/components](https://www.npmjs.com/package/@sooyeoniya/components)
  - `hooks` npm 배포 링크: [@sooyeoniya/hooks](https://www.npmjs.com/package/@sooyeoniya/hooks)
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: [Storybook: Modal Component](https://6812fc18153d0f93012d5a97-cyuwxipvby.chromatic.com/)

## 리뷰 요청 & 논의하고 싶은 내용

## 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

### 💡 조건부 렌더링 vs CSS 조작

모달을 구현하면서 `isOpen`, `title`, `showCloseButton`과 같은 **요소의 표시 여부를 조건부 렌더링으로 처리할지, CSS로 조작할지**에 대한 기준을 고민하게 되었습니다.

```tsx
// 모달 열고 닫는 상태는 CSS로 제어
<Backdrop $isOpen={isOpen} $position={position} onClick={onClose}>
  {/* 생략 */}
</Backdrop>

// 제목과 닫기 버튼은 조건부 렌더링으로 제어
{title && (
  <Title $color={title.color} $size={title.size}>
    {title.text}
  </Title>
)}

{showCloseButton && (
  <CloseButton type="button" onClick={onClose}>
    <IoClose
      color={backgroundColor === "#000" ? "#fff" : "#000"}
      size={30}
    />
  </CloseButton>
)}
```

- `isOpen`은 **모달 전체의 열림, 닫힘 상태**를 담당하기 때문에, 렌더링 여부를 제어하기보다는 **CSS로 시각적 상태만 바꾸는 방식이 더 적절**하다고 판단했습니다.
    - 이 방식은 트랜지션 애니메이션 적용할 때도 편리하고, 내부 구조를 유지하면서 상태만 바꿀 수 있다는 장점이 있다고 생각했습니다.
- 반면 `title`과 `showCloseButton`은 **존재 자체가 선택적인 요소**이기 때문에 **DOM이 꼭 필요하지 않은 경우도 많아서** 조건부 렌더링이 더 자연스럽다고 느꼈습니다.
    - `title`은 단순 문자열 기반이라, 존재 여부를 CSS로 조작하는 것은 오히려 우회적이고 불필요하다고 판단했습니다.
    - `showCloseButton`도 필요하지 않다면 DOM에 포함시키지 않는 편이, **접근성 측면에서도 더 바람직**하다고 생각했습니다.


### 💡 상수 메시지를 함수로 만드는 것에 대한 고민

원래는 유효성 검증 에러 메시지에 대한 상수를 정의하는 것에 대해 아래와 같이 구현하려고 했습니다.

```tsx
// 기존 방식: 상수를 함수로 정의
INVALID_LENGTH: (maxLength: number) => `${maxLength}자리를 입력해주세요.`,
INVALID_YEAR: (year: number) => `유효한 연도(YY)를 입력해주세요.(${year}년 이상)`,
```

위 코드에서는, 값만 바꿔서 다양한 상황에 쓸 수 있어서 함수 형태가 재사용에는 확실히 편하다는 생각이 들었습니다.
하지만, `상수`처럼 보이는데 실제로는 `함수`라서, 사용하는 곳에서는 컨벤션 측면에서 약간 헷갈릴 수도 있겠다는 생각이 들었습니다.

```tsx
// 상수인데 함수처럼 사용하고 있음
ERROR_MESSAGE.INVALID_LENGTH(MAX_LENGTH)
```

그래서 아래처럼 메시지를 고정 문자열로 두고, 필요한 값은 조합하여 사용하도록 변경했습니다!

```tsx
// 고정 문자열로 변경
INVALID_LENGTH: "자리를 입력해주세요.",
INVALID_YEAR: "유효한 연도(YY)를 입력해주세요.",

// 사용 시 조합
`${MAX_LENGTH}${ERROR_MESSAGE.INVALID_LENGTH}`
`${ERROR_MESSAGE.INVALID_YEAR}(${currentYear}년 이상)`
```

## 2) 이번 리뷰를 통해 논의하고 싶은 부분

### 💬 모달 컴포넌트의 유연성과 제약 사이에서의 균형

모달을 구현하면서 **사용자가 어디까지 자유롭게 커스터마이징할 수 있어야 할지**에 대한 고민이 많았습니다.

모달 컴포넌트 라이브러리를 사용하는 목적은 번거로운 기본 기능(e.g., 열기, 닫기, etc.)을 대신 처리해주고, 그 위에 사용자가 자유롭게 커스터마이징할 수 있도록 돕기 위함이라고 생각합니다. 그런데 옵션을 과도하게 제공하거나 스타일과 기능을 지나치게 제한하면, 오히려 사용성이 떨어질 수 있다고 느꼈습니다.

예를 들어, 저희가 구현한 모달에는 `title(제목)`이나 `showCloseButton(닫기 버튼)` 같은 props가 있습니다. 이는 일부 사용자에게는 편리하지만, 다른 사용자에게는 불필요하거나 제약으로 느껴질 수 있습니다. 그래서 이러한 기능은 **선택적으로 제공**하되, **필요할 경우 내부 콘텐츠를 자유롭게 구성**할 수 있도록 열어두는 방식을 택했습니다.

다만 이 과정에서 생긴 고민은, **"기본적인 기능을 얼마나 제공해야 하고, 어디까지 열어줘야 하는가"** 였습니다.

예를 들어, `제목`이나 `닫기 버튼`을 제공하려면 **해당 요소의 스타일링까지 제어 가능하게 해야 할지**도 함께 고려해야 했습니다. 그렇지 않으면 오히려 자유도가 떨어져서 사용자 입장에서는 이를 사용하는 데 더 불편할 것 같다고 생각했어요.

**따라서, 열기, 닫기 같은 핵심 기능만 제어할 수 있도록 하고, 나머지 구조나 UI는 슬롯이나 children을 통해 완전히 커스터마이징할 수 있게 하는 게 더 유연한 설계라고 생각이 듭니다.**

이런 고민의 늪에 빠지다 보니, 결국 props가 이도 저도 아닌 애매한 상태가 되어버린 것 같다는 생각이 들었습니다. 그래서 이 고민을 마르코와 나눠보고 싶었고, 제가 앞서 정한 방향이 괜찮았는지 한번 조언을 듣고 싶습니다.

PS. 만약 핵심 기능만 제어할 수 있도록 한다면, ESC 눌렀을 때 모달 닫히는 것에 대한 유무와, 백드롭 눌렀을 때 모달 닫히는 것에 대한 유무도 제공해주는 것이 일관성 있을까요?

### 💬 공통 유효성 검증 훅 vs 선언적 훅 네이밍

`useCvcValidation`과 `usePasswordValidation` 훅은 `MAX_LENGTH`를 제외하면 거의 동일한 구조입니다.
`useState`로 상태를 만들고, 조건문으로 유효성을 검증하는 부분이 그대로 반복되고 있어서 공통 훅으로 분리할 수 있을 것 같았습니다.

하지만 그렇게 되면 개별 훅이 하는 일이 거의 없어지고, 도메인별 유효성 검사라는 **선언적인 의미**도 사라지게 되더라고요.
훅 이름만 보고 "아, 여기는 CVC나 비밀번호를 검증하겠구나" 하는 식의 **의도를 드러내는 힘**이 줄어드는 게 고민이었습니다.

그래서 생각한 절충안은, **공통 로직은 따로 분리하되**, 실제 사용 시에는 `useCvcValidation`, `usePasswordValidation`처럼 **각 도메인 이름으로 한 번 더 감싸주는** 방식입니다.
이러면 재사용성과 선언적 의미 둘 다 챙길 수 있지 않을까 생각했어요. 이 방향에 대해 마르코의 의견도 궁금합니다!

### 💬 에러 메시지 객체 매핑 vs 명시적인 코드 유지

위와 비슷한 **중복 코드**에 대한 고민인데요, 현재 훅 모듈의 모든 파일에서 아래와 같이 에러 메시지 객체의 구조가 반복된다는 점입니다.

```tsx
{
  isError: true,
  errorMessage: ERROR_MESSAGE.INVALID_NUMBER,
}
```

이런 구조가 여러 곳에서 비슷하게 사용되고 있어서, 아예 객체 매핑으로 만들어두고 `key`만 넘기면 자동으로 반환하게 만들까 고민했습니다. 이렇게 하면 비슷한 구조를 재사용할 수 있기 때문에 코드의 가독성이 좋아진다고 생각했습니다.

다만, 객체 매핑 방식으로 만들 경우 **반환되는 객체가 어떤 것인지 알 수 없어서 오히려 명시성이 떨어질 수도 있다**는 의견이 있었고, 페어랑 논의 끝에 결국 현재 구조를 유지하기로 했습니다!

이 부분에 대해 혹시 마르코라면 어떤 쪽이 더 낫다고 생각하실지 궁금합니다.

### 💬 훅에 값을 인자로 넣는 방식 vs 유효성 검증 함수를 반환하는 방식

현재 구조는 **훅에 값을 직접 인자로 넣고**, 내부에서 유효성 검사를 수행한 뒤 **에러 상태를 반환하는 방식**입니다.

이 방식은 선언적이고 직관적이어서 사용하기 편하다고 느꼈습니다. 사용자는 단순히 값을 넘기기만 하면 되고, 훅 내부에서 자동으로 검증 결과가 갱신되니 간단하게 쓸 수 있다는 점이 장점이라고 생각했어요.

하지만 이 방식은 검증 시점을 사용자가 직접 제어할 수 없고, `useEffect`를 사용하고 있어서 렌더링마다 검증이 반복될 수 있다는 점이 조금 걸렸습니다.

그래서 대안으로, 검증 함수 자체를 반환하는 방식도 고민하게 됐습니다. 검증 함수를 반환해서 사용자가 직접 사용하게 되면, 사용자가 검증 타이밍을 직접 명시할 수 있고, 다양한 이벤트 흐름에 맞게 유연하게 활용 가능하다는 장점이 있다고 생각했습니다.

결론적으로는 두 방식 모두 장단점이 뚜렷해서, 어떤 구조가 더 적절한지는, 실제 사용하는 맥락에 따라 달라질 것 같은데, 이 부분에 대해 마르코의 의견도 궁금합니다!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

wmakerjun committed:
docs: PR 템플릿 개선

sooyeoniya committed:
docs: 초기 문서 작성

- 페어 규칙 및 기능 요구사항 정리

sooyeoniya committed:
feat: 모달 배경 및 창 UI 구현

sooyeoniya committed:
feat: 모달 제목 props 추가

sooyeoniya committed:
feat: 모달 배경 및 제목 색상 props 추가

sooyeoniya committed:
feat: 모달 열림 상태 props 추가

sooyeoniya committed:
feat: 모달 닫힘 상태 props 추가

sooyeoniya committed:
feat: 모달창 위치 props 추가

sooyeoniya committed:
feat: 모달 클릭 시 닫기 이벤트 전파 방지

sooyeoniya committed:
feat: 모달 스토리 종류별로 구현

sooyeoniya committed:
refactor: 공통 args 객체로 분리

sooyeoniya committed:
feat: 모달 제목 글씨 크기 props 추가

sooyeoniya committed:
feat: 모달 닫기 버튼 props 추가

sooyeoniya committed:
chore: 모달 export 추가

sooyeoniya committed:
chore: package.json 패키지명 및 버전 변경

sooyeoniya committed:
chore: npm 공개 패키지 배포 설정 추가

sooyeoniya committed:
feat: global.d.ts에 외부 라이브러리 타입 선언 및 Modal 컴포넌트 적용

sooyeoniya committed:
docs: KPT 회고 작성

sooyeoniya committed:
chore: types 폴더 위치 변경

sooyeoniya committed:
docs: 기능 요구사항 수정

sooyeoniya committed:
feat: ESC 버튼 이벤트 기능 추가

sooyeoniya committed:
chore: npm 버전 패치 v0.0.2

sooyeoniya committed:
chore: npm 버전 0.0.2 적용

sooyeoniya committed:
Merge branch 'woowacourse:main' into step1

sooyeoniya committed:
docs: Hooks 기능 요구 사항 정리

sooyeoniya committed:
feat: 유효성 검증 유틸 함수 구현

- 입력값 길이 검증 함수
- 양의 정수 검증 함수

sooyeoniya committed:
feat: 에러 메시지 상수화

sooyeoniya committed:
feat: 비밀번호 유효성 검증 훅 구현

sooyeoniya committed:
refactor: types 폴더를 src 폴더 내부로 이동

sooyeoniya committed:
feat: CVC 번호 유효성 검증 훅 구현

sooyeoniya committed:
docs: 페어 규칙 및 협업에 대한 정의 작성

- 시지프와 준의 짧은 피드백 강의(20분)를 통해 작성한 내용

sooyeoniya committed:
refactor: useCvcValidation 훅 구조 변경

- useState로 에러 상태 및 에러 메시지 관리
- useEffect로 value값이 변경되었을 때 유효성 검증 및 에러 상태 및 메시지 업데이트

sooyeoniya committed:
refactor: usePasswordValidation 훅 구조 변경

sooyeoniya committed:
feat: useCardNumberValidation 훅 구현

sooyeoniya committed:
refactor: 유효성 검증 시 에러 덮어씌워지지 않도록 return 추가

sooyeoniya committed:
feat: useExpirationDateValidation 훅 구현

sooyeoniya committed:
feat: 빈값 또는 값이 없는 경우 바로 반환 처리

- 이 경우에는 유효성 검사를 하지 않도록 함
- 카드 번호나 만료 기간 같은 경우에는 객체이므로 값이 존재하지 않는 경우도 체크
- 그 외에는 그냥 빈 값인지 체크

sooyeoniya committed:
test: useCardNumberValidation 테스트 구현

sooyeoniya committed:
test: useCvcValidation 테스트 구현

sooyeoniya committed:
test: usePasswordValidation 테스트 구현

sooyeoniya committed:
test: useExpirationDateValidation 테스트 구현

sooyeoniya committed:
test: 카드 번호 유효한 경우 테스트 케이스 추가

sooyeoniya committed:
chore: package.json 패키지명 및 버전 수정

sooyeoniya committed:
chore: 공개 패키지 설정

sooyeoniya committed:
chore: 배포한 hooks 라이브러리 설치

sooyeoniya committed:
feat: createPortal 기능 추가

- document.body로 모달을 걸어주어 z-index 오류 발생하지 않도록 수정

sooyeoniya committed:
refactor: Modal 컴포넌트 전체 스토리 버튼 생성

- Default 컴포넌트를 재사용하여 나머지 스토리 구현

sooyeoniya committed:
chore: 필요 없는 주석 제거

sooyeoniya committed:
chore: npm 버전 0.0.3 적용

sooyeoniya committed:
chore: 배포한 모달 라이브러리 버전 업데이트

sooyeoniya committed:
refactor: 유틸 함수들 하나의 파일에 정의

sooyeoniya committed:
refactor: 공통 유효성 결과값 재사용 객체 생성

sooyeoniya committed:
chore: 훅 npm 버전 업그레이드

sooyeoniya committed:
chore: 버전 의존성 제거

sooyeoniya committed:
fix: 에러가 안날 때 유효성 결괏값 초깃값으로 설정

sooyeoniya committed:
chore: 버전 패치 0.0.4

sooyeoniya committed:
chore: 버전 0.0.4 적용

sooyeoniya committed:
docs: components, hooks README 작성

[추가 작업]
- Modal의 스토리에서 argTypes 추가

sooyeoniya committed:
docs: KPT 회고 작성

sooyeoniya committed:
chore: storybook 배포

sooyeoniya committed:
chore: npm 버전 0.0.4 패치

sooyeoniya committed:
chore: npm 버전 0.0.5 패치 - 훅 라이브러리

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

sooyeoniya commented:
@coderabbitai Based on commit history, identify the recurring work patterns this developer uses to solve problems.

Focus on these two key questions:

What consistent workflow or cycle does the developer follow?

– For example: feature → styling → testing. Explain with specific commit examples.

What does this pattern suggest about the developer’s mindset or strategy?

– e.g., feature-first, cleanup-later, test-driven, etc.

Structure the result with a clear title + example + interpretation, and provide the answer in Korean.

wonsss commented on components/src/lib/Modal/Modal.tsx:
```diff
@@ -0,0 +1,79 @@
+      <ModalBox
+        $backgroundColor={backgroundColor}
+        $position={position}
+        onClick={stopPropagation}
```
👍

wonsss commented on components/src/lib/Modal/Modal.tsx:
```diff
@@ -0,0 +1,79 @@
+          {showCloseButton && (
+            <CloseButton type="button" onClick={onClose}>
+              <IoClose
+                color={backgroundColor === "#000" ? "#fff" : "#000"}
```
backgroundColor 의 타입은 string으로 너무 광범위해요.
색상 타입을 더 구체적으로 정의하는 게 좋을 것 같아요. 그리고 색상 타입을 정의한다면, #000 같은 코드보다는 의미있는 이름의 타입으로 정의하는 게 가독성 등에 좋을 것 같아요.

sooyeoniya replied:
마르코가 언급해주신 부분은, 배경색에 따라 X 아이콘 색상을 두 가지(`#000`, `#fff`)로 분기한 처리였는데요. 이 부분을 객체 상수나 enum처럼 분리해서 관리하면 더 좋겠다는 의미로 이해했습니다!

실제로 배경색(`backgroundColor`)은 사용자가 자유롭게 지정할 수 있도록 `string`으로 타입을 열어두는 것이 확장성 면에서 좋다고 생각했어요. 그래서 해당 prop은 `string` 타입으로 유연하게 설계해뒀었습니다.

다만 말씀하신 부분에서 X 아이콘의 색상은 `#000`, `#fff`로만 제한하고 싶었기 때문에 이 부분은 따로 상수로 분리해 관리했습니다.

여기서 한 가지 고민이 되었던 점은, 이번 미션의 요구사항에서는 배경색이 기본적으로 흰색으로 고정되어 있어서, `backgroundColor`를 열어둔 설계 자체가 현재로서는 과한 처리일 수도 있다는 점이었습니다. 특히 2단계 요구사항을 보면 버튼도 흰색과 검정색 두 가지 버전으로 제한되어 있기 때문에, 아예 `light` / `dark` 같은 **모드 기반의 테마 설정**으로 방향을 바꾸는 게 더 명확하고 타입 안정성도 높을 것 같다는 생각이 들었습니다.

그래서 이번에는 색상을 두 가지로 고정하고, `mode` 기반으로 처리하는 방식으로 리팩토링을 시도해봤습니다! https://github.com/woowacourse/react-modules/pull/105/commits/a0b4660fffcd452d662d218eb4f0c869d39fce5e

wonsss commented on components/src/lib/Modal/Modal.tsx:
> 💬 모달 컴포넌트의 유연성과 제약 사이에서의 균형
모달을 구현하면서 사용자가 어디까지 자유롭게 커스터마이징할 수 있어야 할지에 대한 고민이 많았습니다.
모달 컴포넌트 라이브러리를 사용하는 목적은 번거로운 기본 기능(e.g., 열기, 닫기, etc.)을 대신 처리해주고, 그 위에 사용자가 자유롭게 커스터마이징할 수 있도록 돕기 위함이라고 생각합니다. 그런데 옵션을 과도하게 제공하거나 스타일과 기능을 지나치게 제한하면, 오히려 사용성이 떨어질 수 있다고 느꼈습니다.
예를 들어, 저희가 구현한 모달에는 title(제목)이나 showCloseButton(닫기 버튼) 같은 props가 있습니다. 이는 일부 사용자에게는 편리하지만, 다른 사용자에게는 불필요하거나 제약으로 느껴질 수 있습니다. 그래서 이러한 기능은 선택적으로 제공하되, 필요할 경우 내부 콘텐츠를 자유롭게 구성할 수 있도록 열어두는 방식을 택했습니다.
다만 이 과정에서 생긴 고민은, "기본적인 기능을 얼마나 제공해야 하고, 어디까지 열어줘야 하는가" 였습니다.
예를 들어, 제목이나 닫기 버튼을 제공하려면 해당 요소의 스타일링까지 제어 가능하게 해야 할지도 함께 고려해야 했습니다. 그렇지 않으면 오히려 자유도가 떨어져서 사용자 입장에서는 이를 사용하는 데 더 불편할 것 같다고 생각했어요.
따라서, 열기, 닫기 같은 핵심 기능만 제어할 수 있도록 하고, 나머지 구조나 UI는 슬롯이나 children을 통해 완전히 커스터마이징할 수 있게 하는 게 더 유연한 설계라고 생각이 듭니다.
이런 고민의 늪에 빠지다 보니, 결국 props가 이도 저도 아닌 애매한 상태가 되어버린 것 같다는 생각이 들었습니다. 그래서 이 고민을 마르코와 나눠보고 싶었고, 제가 앞서 정한 방향이 괜찮았는지 한번 조언을 듣고 싶습니다.
PS. 만약 핵심 기능만 제어할 수 있도록 한다면, ESC 눌렀을 때 모달 닫히는 것에 대한 유무와, 백드롭 눌렀을 때 모달 닫히는 것에 대한 유무도 제공해주는 것이 일관성 있을까요?

정말 좋은 고민입니다! 
모달을 구현할 때는 우선 기본 기능만 갖춘 Primitive Modal부터 만드는 것이 좋다고 생각해요.
이렇게 만든 모달은 어떤 경우에도 재사용 가능한 가장 유연하고 추상화된 컴포넌트가 됩니다.

그런데 실제로 모달을 사용하는 상황을 보면, 단순히 열고 닫는 것 외에도
타이틀, 설명, 닫기 버튼 같은 UI 요소들이 함께 필요한 경우가 많고,
이런 UI들에 대해 일관된 스타일이나 기능을 적용하고 싶어지죠.

이럴 땐, 해당 역할을 담당하는 세부 컴포넌트들을 함께 제공하는 게 좋아요.
예를 들어 Modal.Header, Modal.Title, Modal.Description 같은 식으로요.

이렇게 구성하면 사용자는 각 UI를 필요한 만큼 조합해서 사용할 수 있기 때문에
유연하면서도 편리한 구조가 됩니다. 이런 설계 패턴을 [Compound Component 패턴](https://patterns-dev-kr.github.io/design-patterns/compound-pattern/)이라고 합니다.

예시를 들면 다음과 같습니다!

```ts
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog"

<Dialog>
  <DialogTrigger>Open</DialogTrigger>
  <DialogContent>
    <DialogHeader>
      <DialogTitle>Are you absolutely sure?</DialogTitle>
      <DialogDescription>
        This action cannot be undone. This will permanently delete your account
        and remove your data from our servers.
      </DialogDescription>
    </DialogHeader>
  </DialogContent>
</Dialog>
```

이런 방식은 모달의 핵심 기능은 Dialog와 DialogContent가 책임지고,
표현(타이틀, 설명 등)은 세부 컴포넌트로 분리되어 있어서
스타일링이나 기능을 개별적으로 조절하기도 좋고, 유지보수 측면에서도 장점이 많습니다.

sooyeoniya replied:
확실히 컴파운드 컴포넌트 패턴은 사용자 입장에서 더 다양한 방식으로 컴포넌트를 조합해서 사용할 수 있다는 큰 장점이 있는 것 같습니다!

이 방식을 고민하면서 느낀 점은, “사용자에게 어디까지 조합할 수 있게 해야 할까?”가 가장 큰 고민인 것 같아요. 예를 들어, 백드롭과 모달 컨테이너는 사용자가 직접 조합해서 쓰기보다는 하나의 모달로 제공하는 것이 훨씬 더 유용하다고 생각합니다. 따라서, 이 방식을 사용하게 되면, 조합을 더 다양하게 했을 때 발생할 수 있는 UI/UX 관련 이슈들을 모두 해결해야한다는 점에서 고민이 많이 되네요 ㅎㅎ

물론, 2단계 미션에서는 컴파운드 컴포넌트 패턴이 더 유용할 것 같다는 생각이 듭니다! 여러 가지 종류의 모달을 만들어야 하는데, 이런 종류들은 컴포넌트를 조합해서 내보내면 훨씬 간편하고, 또 이런 종류 외에도 사용자가 원하는 다양한 방식으로 모달을 커스터마이징 할 수 있도록 하면 좋을 것 같습니다. 

다만, 모든 것을 조합해서 사용하는 것보다는, 어떤 것이 props로 제공되면 좋을지, 어떤 부분이 컴포넌트로 분리하면 좋을지 잘 구분해서 구현하면 훨씬 더 효율적일 것 같다는 생각이 듭니다!

우선 지금 방식으로 구현을 진행한 후, **컴파운드 컴포넌트 패턴** 방식만의 장점이 더 부각된다면, 그때 리팩토링을 고려해보는 방향으로 진행하겠습니다! 좋은 정보를 주셔서 정말 감사합니다. 😊

wonsss replied:
잘 이해해주셨네요! 컴파운드 컴포넌트 패턴의 유용성이나 한계점은  앞으로 직접 겪어보며 느끼시면 더 잘 이해되실 거예요!

wonsss commented on components/src/lib/Modal/Modal.tsx:
```diff
@@ -0,0 +1,79 @@
+    e.stopPropagation();
+  };
+
+  useEffect(() => {
+    const handleEscape = (e: KeyboardEvent) => {
+      if (e.key === "Escape") onClose();
+    };
+
+    document.addEventListener("keydown", handleEscape);
+    return () => document.removeEventListener("keydown", handleEscape);
+  }, [onClose]);
```
커스텀훅으로 분리하면 어떨까요?

그리고 현재 키보드 이벤트 처리를 통한 키보드 접근성 향상이 Escape 키만 있는데, 일반적으로 모달에 기대되는 사용성 등은 다음과 같습니다.

참고만 해주셔도 되고, 추가 구현해보실 수도 있을 것 같아요.

- 포커스 관리
모달이 열릴 때 자동으로 모달 내부로 포커스되어야 합니다.
모달이 닫힐 때 이전 포커스 위치로 돌아가야 합니다.

- Tab 키 관리
Tab 키로 모달 내부 요소들 간 이동할 수 있고 내부 요소들 이동으로 제한할 수 있어야 합니다.
Enter 키나 Space 키로 버튼 클릭이 처리되어야 합니다.


W3에서 제공하는 Modal Dialog 예시를 직접 실행해보시면, 위 설명을 이해하는 데 도움이 되실거예요.

https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/

sooyeoniya replied:
ESC 키보드 이벤트 관련 로직을 커스텀 훅으로 분리해두었습니다. 확실히 관심사 분리도 잘 되고, 코드가 깔끔해져서 좋네요! ㅎㅎ https://github.com/woowacourse/react-modules/pull/105/commits/2594f618b7c7b8637f4feaf37963f9282adf4f9c

말씀해주신 포커스 트랩과 Tab 키 관리의 경우에는 2단계 요구사항에도 존재하는 부분인데요, 아래 스크린 리더와 함께 2단계에서 한 번 더 고민해보면서 반영해보겠습니다! 의견 주셔서 정말 감사해요 😄

wonsss replied:
감사합니다!

wonsss commented on components/src/lib/Modal/Modal.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  return createPortal(
+    <Backdrop $isOpen={isOpen} $position={position} onClick={onClose}>
+      <ModalBox
```
모달 컴포넌트들간에 접근성 속성을 추가해 모달의 역할과 관계를 명확히 하면 좋을 것 같아요.

이렇게 수정하면 스크린 리더 사용자가 모달이 열리고 닫혔고, 모달의 제목, 내용 등을 이해할 수 있어요.

참고
ex) role="dialog", aria-modal="true", aria-labelledby, aria-describedby, aria-hidden
[WAI-ARIA Dialog Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/)

[ARIA Dialog Role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/dialog_role)

sooyeoniya replied:
웹 접근성(A11y)에 대해서는 앞서 말씀드린 바와 같이 step2 요구사항이기도 한데요! 물론, 스크린 리더에 대한 접근성은 명시적으로 요구사항에 존재하지는 않지만 그래도 한 번 추가해봤습니다!

자세한 내용은 커밋 내용에도 남겨두었습니다. https://github.com/woowacourse/react-modules/pull/105/commits/505f51f653fada68c5947ff84689aeb0bc5d21f6

커밋 내용을 간략히 요약하자면, 

- 백드롭, 모달에 대한 role을 정의해주었구요.
- 모달 제목과 내용 부분을 aria-labelledby, aria-describedby를 통해 id로 지정해주었습니다.

아직 해결하지 못한 부분도 있습니다! 여러 테스트를 진행했으나 아래 이슈들에 대해서는 현재로서는 제대로 해결되지 않았습니다. 다만, 우선순위가 급하지 않다고 판단하여 다른 기능을 먼저 구현하고 다시 고민해볼 예정입니다. 따라서, 본 이슈는 커밋 내용에 따로 남겨두었습니다.

실제 MacOS의 내장 스크린 리더인 VoiceOver를 사용하여 테스트한 결과, 다음과 같은 이슈가 발생했습니다.

- Title이 한글일 때 제대로 읽히지 않는 현상 (영어는 정상적으로 읽힘)
- CloseButton이 잘못 읽히고 있음
    - 예상: “close, 버튼”
    - 실제: “이미지, 버튼, 그룹”

wonsss commented on components/src/lib/Modal/Modal.stories.tsx:
스토리북의 상호작용 테스트를 활용해보시면 좋을 것 같아요. 실제 사용자의 행동을 시뮬레이션하여 컴포넌트가 예상대로 동작하는지 확인하는 데 매우 유용합니다.
테스트코드 play 함수를 스토리 객체에 추가하면, 스토리북에서 해당 컴포넌트의 상호작용을 자동으로 테스트할 수 있어요.

참고: https://storybook.js.org/docs/writing-tests/component-testing

```ts
export const MyStory: Story = {
  // 1. 렌더링 함수
  render: () => <MyComponent />,
  
  // 2. 상호작용 테스트
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    
    await step('테스트 단계 설명', async () => {
      // 테스트 코드
    });
  }
};
```

sooyeoniya replied:
사실 스토리북 상호작용 테스트에 대해서 알고 있었는데, 이번에 한 번 직접 구현해보고자 하여 모달에 대한 상호작용 테스트를 총 2가지 작성해봤어요. https://github.com/woowacourse/react-modules/pull/105/commits/45114e711d9f8d0d6da4658845b648f12b9b9d0e

1. 모달 열고 닫는 기능 테스트
2. 모달의 ESC 키보드 이벤트 테스트

기존에는 `isOpen`이 `false`일 때, 단순히 `display: none`을 사용해 모달을 화면에 보이지 않게 처리하고 있었는데요. 이 방식은 화면에는 보이지 않더라도 DOM에는 여전히 요소가 존재하기 때문에, 모달이 완전히 사라졌는지를 확인해야 하는 테스트에서 오류가 발생했습니다.

그래서 조건부 렌더링을 통해 아예 DOM에서 모달 요소 자체를 제거하는 방식으로 변경했습니다!

이렇게 하면 보이지 않는 불필요한 DOM 요소를 제거할 수 있어 메모리 측면에서도 효율적이고, 스크린 리더 입장에서도 존재하지 않는 요소는 완전히 무시할 수 있기 때문에 접근성 측면에서도 더 나은 방식이라고 생각했습니다. 그리고 앞서 말씀드린 것처럼 테스트에서도 예상한 대로 명확히 동작하기 때문에 테스트의 신뢰성과 유지보수 측면에서도 더 유리한 것 같습니다 :)

wonsss reviewed (changes requested):
안녕하세요 니야 😄 
다시 만나서 반갑습니다~! 페이먼츠 모듈 미션하시느라 수고 많으셨어요.
이번 미션에서도 좋은 고민을 많이 하신 것 같아 좋네요! 👍 
리뷰 확인 후 재리뷰 요청주세요. 🙇

sooyeoniya committed:
remove: 불필요한 파일들 제거

- 라이브러리용 패키지이므로 웹 애플리케이션용 파일들 제거
- 테스트를 위해 설치해둔 내 라이브러리 의존성 제거

sooyeoniya committed:
chore: 빌드 시 라이브러리에서 타입 정의 파일 생성

sooyeoniya committed:
chore: hooks 모듈 npm 버전 업데이트 0.0.6

sooyeoniya committed:
chore: components 모듈 빌드 시 타입 정의 파일 생성

sooyeoniya committed:
refactor: 카드 번호 훅 역할 수정

- 입력 필드의 상태 관리 + 유효성 검사 로직을 함께 포함하는 입력 단위의 훅으로 변경

sooyeoniya committed:
refactor: 유효 기간 훅 역할 수정

- 입력 필드의 상태 관리 + 유효성 검사 로직을 함께 포함하는 입력 단위의 훅으로 변경

sooyeoniya committed:
refactor: CVC 번호 훅 역할 수정

- 입력 필드의 상태 관리 + 유효성 검사 로직을 함께 포함하는 입력 단위의 훅으로 변경

sooyeoniya committed:
refactor: 비밀번호 훅 역할 수정

- 입력 필드의 상태 관리 + 유효성 검사 로직을 함께 포함하는 입력 단위의 훅으로 변경

sooyeoniya committed:
refactor: 커스텀 훅 공통 로직 분리

- 공통으로 유효성 검사하는 '빈 값', '양의 정수', '입력 길이'에 대한 로직 공통 함수로 분리

sooyeoniya committed:
test: 커스텀 훅 로직 변경에 맞춰 테스트 코드 전체 수정

sooyeoniya committed:
docs: 커스텀 훅 README 수정

sooyeoniya committed:
chore: 커스텀 훅 버전 업데이트 0.0.7

sooyeoniya committed:
feat: 모달 light/dark 테마 적용

sooyeoniya committed:
refactor: ESC 키보드 이벤트 커스텀 훅으로 분리

sooyeoniya commented:
@coderabbitai pause

sooyeoniya committed:
feat: 모달 컴포넌트 각 role 및 ARIA 속성 추가

[반영 사항]
- 백드롭과 모달에 대한 role 정의
- 모달은 dialog 역할
- 백드롭은 presentation 역할 부여
- 백드롭의 경우 의미없는 요소이므로 시맨틱 역할 제거하도록 함
- 단 aria-hidden="true"가 아닌 role="presentation"을 사용하여, 해당 요소 자체만 무시되고, 자식 콘텐츠는 스크린 리더가 읽도록 수정

- useId 훅을 사용해 새 식별자 생성하여 title과 content에 대한 ID 생성
- aria-labelledby : 모달 제목 지정
- aria-describedby : 모달 내용 지정

- IoClose (React Icon 닫기 버튼)의 경우 svg이므로 읽히지 않고 포커스 무시되도록 설정함 -> 자꾸 포커스가 내부에 또 잡혀서 "이미지" 이렇게 한 번 더 읽힘

[해결하지 못한 문제 (1)]
MacOS에 내장되어 있는 VoiceOver를 통해 테스트해본 결과,
title이 한글일 때는 제대로 안읽힐 때가 많음.. 왜일까?
"Modal 모달 Modal" -> 이렇게 되어 있으면 영어만 읽고, 한국어는 그냥 스킵됨
-> 이때 음성 자체가 다른 성우(영어 전문 성우)로 바뀜. 해결 방법 찾지 못함.

[해결하지 못한 문제 (2)]
CloseButton 이 부분 자체가 "이미지, 버튼, 그룹" 이렇게 읽힘
-> 내부 로직에 따르면 "close, 버튼" 이렇게 읽혀야 정상인데,
이것 저것 시도해봐도 바뀌지 않아서 도대체 어떤 부분이 문제인지 아직 파악 불가.

sooyeoniya commented on hooks/src/App.tsx:
이 파일이 남아 있었던 이유는, 데모 페이지를 어떻게 다뤄야 할지 고민했던 부분이 있어서였습니다. `components` 모듈은 Storybook을 통해 충분히 확인할 수 있었지만, `hooks`는 실제 사용 예시를 보려면 직접 코드를 작성해야 하더라고요. 그래서 별도로 데모 페이지를 제공하는 게 좋을지 한참 고민했습니다. 다만 이번 미션의 요구사항을 고려했을 때, 데모 페이지 구현은 우선순위가 높지 않다고 판단해서 일단은 남겨둔 채 진행했고, 그 결과 `App.tsx` 파일이 그대로 남아 있게 된 것 같아요. 지금은 이 파일을 정리하고 삭제하는 게 더 맞겠다는 판단이 들어서 그렇게 진행하려고 합니다! https://github.com/woowacourse/react-modules/pull/105/commits/c2dd56428993ac30565c95b2bdf650f1b49480f2

다만, 한 가지 조금 헷갈리는 부분이 있었는데요.
이번 미션의 PR 템플릿을 보면 데모 페이지에 대한 링크를 작성하는 항목이 포함되어 있더라구요.

<img width="706" alt="image" src="https://github.com/user-attachments/assets/699027d5-691e-4f4c-b615-a65ff75d7f6c" />


이 부분을 보고, 라이브러리 내부에서 `App.tsx`를 활용해 간단한 데모 페이지를 직접 만들어두는 것이 일반적으로도 자주 쓰이는 방식인지 궁금했습니다. 혹시 이 부분에 대해서 경험적으로 조언 주실 수 있을까요?

wonsss replied:
실제 배포 라이브러리 코드와는 별도로, examples이나 demo 같은 디렉터리를 만들어 데모 앱을 구성하곤 합니다.
데모 페이지는 라이브러리 패키지(dist/)에 포함하지 않도록 하여, 패키지 크기에 영향주지 않도록 해야 할 것 같아요.

sooyeoniya commented on hooks/src/types/global.d.ts:
```diff
@@ -0,0 +1 @@
+declare module "@sooyeoniya/hooks";
```
라이브러리를 개발하는 과정에서, 본 프로젝트 내에서 해당 라이브러리를 직접 설치해, 사용 흐름을 테스트하던 중, 타입 추론을 위해 임시로 추가해두었던 코드입니다. 하지만 라이브러리 자체에는 불필요한 부분이므로 제거해주었습니다!

외부에서 `@sooyeoniya/hooks` 패키지를 설치하여 사용할 때, 타입이 `any`로 추론되는 문제를 해결하기 위해 `declare module` 을 통해 수동으로 타입을 정의해주어야 했는데요, 이렇게 수동으로 작성하는 방식은 생산성 측면에서 비효율적이라고 판단했습니다. 따라서, 라이브러리를 사용 할 때 타입 정보가 자동으로 추론될 수 있도록, 빌드 시 `dist/` 디렉터리에 타입 정의 파일도 함께 생성되도록 설정하였습니다. https://github.com/woowacourse/react-modules/pull/105/commits/9e89f30809dbcb239e96e407534e1fb3e19b6105

wonsss replied:
잘 수정해주셨네요!

sooyeoniya commented on hooks/src/lib/hooks/useCardNumberValidation.ts:
```diff
@@ -0,0 +1,68 @@
+  const [cardNumberValidationResult, setCardNumberValidationResult] =
+    useState<CardNumberValidationType>(defaultCardNumberValidationValue);
+
+  useEffect(() => {
```
이번 미션 요구사항에 맞게 입력 필드의 상태 관리를 포함하여 커스텀 훅을 모두 수정했습니다!

useCardNumber 훅에 대한 커밋입니다! https://github.com/woowacourse/react-modules/pull/105/commits/f1b5eb3d2cda6f2ac638b84491beab201f223a5b (다른 훅에 대한 커밋은 남겨주신 각 코멘트에 모두 남겨두었습니다!)

추가 동시에 작업한 내용으로 [테스트 코드 수정](https://github.com/woowacourse/react-modules/pull/105/commits/5f9dddbddde56a6aa000a7478d72e78c234e3703)과 [리드미 수정](https://github.com/woowacourse/react-modules/pull/105/commits/bbd28f5b2ef6844ef9a3fa8a08e247754b246a7f) 커밋입니다 😃

리드미에서 확인하실 수 있다시피 모든 훅에는 입력값과, 입력값 변경을 처리하는 핸들러 함수와, 유효성 검증 결과를 담은 객체 이렇게 총 3개를 반환할 수 있도록 해주었습니다.

핸들러 함수의 경우에는 모두 이벤트 객체가 아닌 value를 받도록 해주었는데, 이벤트 타입을 직접 `ChangeEvent<HTMLInputElement>`로 고정하는 것보다 사용자가 더 다양한 이벤트 상황에서 해당 핸들러를 사용할 수 있도록 확장성을 고려하여 단순 value만 받을 수 있도록 해주었습니다. 카드 번호와 유효 기간의 경우에는 해당 필드가 어느 필드인지도 그 위치를 받을 수 있도록 해주었어요!

wonsss replied:
잘 수정해주셨네요!

sooyeoniya commented on hooks/src/lib/hooks/useCvcValidation.ts:
```diff
@@ -0,0 +1,36 @@
+
+const MAX_LENGTH = 3;
+
+const useCvcValidation = (value: string): ValidationType => {
```
useCvcNumber 훅도 동일하게 수정하였습니다. https://github.com/woowacourse/react-modules/pull/105/commits/df525a82fd01cee93ee6a69251a20de7469452a7

sooyeoniya commented on hooks/src/lib/hooks/useExpirationDateValidation.ts:
```diff
@@ -0,0 +1,91 @@
+
+const useExpirationDateValidation = (
+  args: ExpirationDateArgs = {}
+): ExpirationDateValidationType => {
```
useExpirationDate 훅도 동일하게 수정하였습니다. https://github.com/woowacourse/react-modules/pull/105/commits/471057f34799e60c5b9cfd558971be080242d3d4

sooyeoniya commented on hooks/src/lib/hooks/usePasswordValidation.ts:
```diff
@@ -0,0 +1,36 @@
+
+const MAX_LENGTH = 2;
+
+const usePasswordValidation = (value: string): ValidationType => {
```
usePassword 훅도 동일하게 수정하였습니다. https://github.com/woowacourse/react-modules/pull/105/commits/e99ece1035061f92bf5e67442fb5ade6deae8906

sooyeoniya commented on hooks/src/lib/hooks/useCardNumberValidation.ts:
```diff
@@ -0,0 +1,68 @@
+          ...prev,
+          [key]: {
+            isError: true,
+            errorMessage: ERROR_MESSAGE.INVALID_NUMBER,
```
마르코 의견처럼 에러 관련 객체만 따로 매핑하는 방식은 가독성이 떨어져서 반영하지 않았구요, 
아예 `validateNumberFieldWithLength` 라는 공통 유효성 검사 함수를 따로 분리하여 적용해주었습니다. 
반환 타입도 명시해주고 있어서 타입 체킹에 문제도 없을 것 같고, 각 훅의 내부 로직이 훨씬 간결해진 장점이 있는 것 같습니다! https://github.com/woowacourse/react-modules/pull/105/commits/3c739693e8e2084ccd81d48ff8aa5f2fba012637

sooyeoniya commented on components/src/lib/Modal/Modal.tsx:
```diff
@@ -0,0 +1,79 @@
+    return () => document.removeEventListener("keydown", handleEscape);
+  }, [onClose]);
+
+  return createPortal(
```
좋은 질문 감사합니다 😄

제가 `createPortal`을 사용하면서 느꼈던 이점들과, 추가로 찾아본 이점들에 대해 정리해보았습니다!

우선 `createPortal`를 도입한 가장 큰 이유는 z-index 관리 측면이었습니다. 포탈을 사용하면 원하는 요소를 특정 DOM 계층이 아닌, 최상위 계층인 `body`에 직접 렌더링할 수 있어서(`document.body`에 걸어주는 방식으로 처리했습니다), z-index 충돌 없이 안정적으로 모달을 띄울 수 있더라구요.

그리고 모달이 부모 컴포넌트 내부가 아니라 `body` 바로 아래에 렌더링되다 보니, 부모 요소의 CSS 스타일에 영향을 받지 않는다는 점도 큰 장점이었습니다.

실질적으로 위 문제들을 해결하기 위해 `createPortal`를 도입하게 되었던 것 같습니다.

추가로 찾아본 장점은 다음과 같습니다!

- **스크린 리더 접근성**
    - 모달이 DOM 최상위에 위치하기 때문에 스크린 리더에 더 명확하게 인식됩니다. 보통 스크린 리더는 DOM 구조를 따라서 내부 콘텐츠를 읽게 되는데, `createPortal`을 사용하면 모달이 `body`의 마지막 자식 요소에 위치하니까, 스크린 리더가 모달을 더 명확하게 인식해서 **"지금 중요한 UI가 떴다"** 는 걸 인지하기 쉬워진다고 합니다.
- **키보드 네비게이션 (포커스 트랩)**
    - 모달은 일반적으로 화면에서 가장 위에 떠야 하므로 `z-index`가 높게 설정됩니다. `createPortal`을 사용하면 모달이 다른 컴포넌트와 DOM 상으로도 분리되기 때문에, 외부 요소에 포커스가 가는 걸 막는 관리가 쉬워진다고 합니다. 예를 들어서, Tab 키를 눌렀을 때 포커스가 모달 내부에서만 순환되도록 쉽게 처리할 수 있어, UX 측면에서도 더 좋은 경험을 제공할 수 있다고 생각합니다.
- **이벤트 버블링 정상 작동**
    - `createPortal`로 모달이 `body` 바로 하위에 렌더링된다고 해도, React 입장에서는 여전히 현재 해당 모달을 호출한 부모 컴포넌트 안에 있는 자식이기 때문에, 이벤트 전달이나 props 같은 기능은 원래처럼 잘 작동합니다. 따라서, 부모에서 넘긴 `onClose` 같은 props나 콜백 함수가 잘 전달되고, React의 이벤트 버블링도 컴포넌트 트리 기준으로 작동하기 때문에, 모달 내부 이벤트도 부모까지 잘 전달됩니다!
- **컴포넌트 위치 독립성**
    - 컴포넌트 트리 어디에서든 모달을 호출해도 일관되게 렌더링이 된다는 장점이 있습니다.

wonsss replied:
꼼꼼하게 잘 정리해주셨네요! 감사합니다

sooyeoniya committed:
test: 모달 interaction 테스트 구현

- 모달 열고 닫는 기능 테스트
- 모달 ESC 키보드 이벤트 테스트

[추가 수정]
- 모달의 isOpen 상태가 false인 경우 아예 DOM에서 완전 제거되도록 수정
-> 이렇게 하면 모달 interaction 테스트할 때도 DOM에 존재하지 않는 것을 명시적으로 확인 가능

sooyeoniya commented:
안녕하세요 마르코! 니야입니다.
개인 일정으로 인해 리뷰 반영이 예상보다 많이 늦어졌습니다. 기다리게 해서 죄송해요 😢 다음부터는 일정이 지연될 경우, 미리 말씀드리겠습니다!

전반적으로 남겨주신 리뷰 하나하나 정말 좋았고, 덕분에 깊이 있게 고민하고 공부할 수 있었던 것 같아요. step1 하면서 궁금했던 점에 대해서도 열심히 답변 달아주셔서 진심으로 감사드립니다 😊
각 코멘트에 모두 답변 달아두었으니 확인 부탁드릴게요! 혹시 이상하거나 잘못된 부분이 있다면 편하게 말씀해주세요 🙇🏻‍♀️

wonsss reviewed (approved):
안녕하세요! 니야 😄 
리뷰 반영 잘 해주셨고, 답변도 상세히 남겨주셔서 감사합니다. 👍 
꼼꼼하게 모든 리뷰 다 잘 반영해주셨고, 1단계 목표에 충분한 것 같아 여기서 merge할게요.
2단계에서 뵈어요! 화이팅입니다 💪

sooyeoniya commented:
@coderabbitai resume

sooyeoniya commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

sooyeoniya commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

sooyeoniya commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

sooyeoniya commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.

sooyeoniya commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, evaluate how well it aligns with the accessibility (A11y) principles of modern web apps.  
Specifically, check for proper use of semantic HTML (like <button>, <nav>, <header>), ARIA attributes for assistive technologies, and proper support for keyboard navigation and focus management.  
If the code already meets these principles effectively, affirm that with a short explanation.  
If it could be improved, suggest one possible direction—but only if the improvement adds clear value without overengineering.

# How to respond  
Start your response with the following markdown title:  
# ✨ 접근성 원칙에 따른 평가 및 개선 포인트

Then respond with:
- ✅ **현재 코드가 원칙에 잘 맞는 이유** (if applicable)  
- ⚠️ **개선 여지가 있다면, 가능한 방향** (optional)  
- **AS-IS**: _Show the current code_
- **TO-BE**: _Show a refactored example_
- **스스로 점검할 수 있는 질문 하나**

# Style  
Use bullet points only. Do not force a suggestion—only provide one if it adds meaningful value.

# Language  
Please write your response **in Korean**.
